packages/agreement-process/src/services/agreementService.ts

```diff
@@ -277,3 +346,160 @@ export async function updateAgreementLogic({
+  );
+  if (newDescriptor === undefined) {
+    throw publishedDescriptorNotFound(agreementToBeUpgraded.data.eserviceId);
+  }
+  const latestDescriptorVersion = z
+    .preprocess((x) => Number(x), z.number())
+    .safeParse(newDescriptor.version);
+  if (!latestDescriptorVersion.success) {
+    throw unexpectedVersionFormat(eservice.data.id, newDescriptor.id);
+  }
+
+  const currentDescriptor = eservice.data.descriptors.find(
+    (d) => d.id === agreementToBeUpgraded.data.descriptorId
+  );
+  if (currentDescriptor === undefined) {
+    throw descriptorNotFound(
+      eservice.data.id,
+      agreementToBeUpgraded.data.descriptorId
+    );
+  }
+
+  const currentVersion = z
+    .preprocess((x) => Number(x), z.number())
+    .safeParse(currentDescriptor.version);
+  if (!currentVersion.success) {
+    throw unexpectedVersionFormat(eservice.data.id, currentDescriptor.id);
+  }
+
+  if (latestDescriptorVersion.data <= currentVersion.data) {
+    throw noNewerDescriptor(eservice.data.id, currentDescriptor.id);
+  }
+
+  validateCertifiedAttributes(newDescriptor, tenant.data);
+
+  const verifiedValid = verifiedAttributesSatisfied(
+    agreementToBeUpgraded.data.producerId,
+    newDescriptor,
+    tenant.data
+  );
+
+  const declaredValid = declaredAttributesSatisfied(newDescriptor, tenant.data);
+
+  if (verifiedValid && declaredValid) {
+    // upgradeAgreement
+    const stamp: AgreementStamp = {
+      who: authData.organizationId,
+      when: new Date(),
+    };
+    const deactivated: Agreement = {
+      ...agreementToBeUpgraded.data,
+      state: agreementState.archived,
+      stamps: {
+        ...agreementToBeUpgraded.data.stamps,
+        archiving: stamp,
+      },
+    };
+    const upgraded: Agreement = {
+      ...agreementToBeUpgraded.data,
+      id: uuidv4(),
+      descriptorId: newDescriptor.id,
+      createdAt: new Date(),
+      updatedAt: undefined,
+      rejectionReason: undefined,
+      stamps: {
+        ...agreementToBeUpgraded.data.stamps,
+        upgrade: stamp,
+      },
+    };
+
+    return [
+      toCreateEventAgreementUpdated(
+        deactivated,
+        agreementToBeUpgraded.metadata.version
+      ),
+      toCreateEventAgreementAdded(upgraded),
+    ];
+  } else {
+    // createNewDraftAgreement
+    await verifyConflictingAgreements(
+      agreementToBeUpgraded.data.consumerId,
+      agreementToBeUpgraded.data.eserviceId,
+      [agreementState.draft],
+      agreementQuery
+    );
+    const createEvent = await createAgreementLogic(
+      {
+        eserviceId: agreementToBeUpgraded.data.eserviceId,
+        descriptorId: newDescriptor.id,
+      },
+      authData,
+      agreementQuery,
+      eserviceQuery,
+      tenantQuery
+    );
+
+    const docEvents = await createAndCopyDocumentsForClonedAgreement(
+      createEvent.streamId,
+      agreementToBeUpgraded.data,
+      1,
+      fileCopy
+    );
+
+    return [createEvent, ...docEvents];
```

Comment:
extract to new function `createNewDraftAgreement` ?

Reply:
In the old codebase the `createNewDraftAgreement` is defined inside the scope of the `agreement upgrade` function so I think that it's not used by other service logics.

Reply:
Perfect! if not also required this is the right place 👍🏻